### PR TITLE
Gripper

### PR DESCRIPTION
**Pull request by bayesian**
changes: @Jialn @emailweixu @runjerry 

1. control each joint honoring limits (otherwise gripper joint wouldn't open, as its limit is over 10x larger)
2. correct way of implementing contact sensor in pygazebo
3. switch pr2 model.config to base on grocery_ground_pr2 branch, but modify the joint settings related to r_arms.
4. modify friction settings
5. add short tables (shorter than standard one in gazebo)
6. switch from camera-based to internal state based.

With the new env, we could successfully grasp successfully after 4-5M environment steps with tf_agent/ppo

### CODE REVIEW

File: python/social_bot/envs/pr2.py [hunk 1]
```diff
@@ -58,128 +81,288 @@ def __init__(self,
...
+
+        if delta_reward > 0:
+            reward += delta_reward
+        else:
+            reward += -0.01     #if no positive reward, penalize it a bit to speed up
+
+        if self._motion_loss > 0.0:
+            v2s = 0.0
+            for joint in self._r_arm_joints:
+                v = self._agent.get_joint_state(joint).get_velocities()[0]
```
Comment (Jialn) [inline python/social_bot/envs/pr2.py]: Maybe we can also simply use actions here. Naturally a human may want to avoid extra efforts (forces of the joints).

File: python/social_bot/envs/pr2.py [hunk 2]
```diff
@@ -58,128 +81,288 @@ def __init__(self,
...
         self._prev_dist = self._get_finger_tip_distance()
-        return self._get_observation()
+        self._prev_gripper_pos = self._get_gripper_pos()
+        self._gripper_reward_dir = 1
+        return obs
 
     def _move_goal(self):
-        loc = (0.6 + 0.3 * (random.random() - 1), 0.4 * (random.random() - 1),
-               0)
+        loc = (0.76 + 0.14 * (random.random() - 1),  0.1 * (random.random() - 1.0),
```
[the remark this answers is not in the document]
Comment (bayesian) [reply python/social_bot/envs/pr2.py]: The PR2 robot is at (0, 0), and this base is pretty big. The table is (0.9, 0), since we only control the right arm, so we put the goal object on the right side (0.76 + 0.14 (random - 0.5)).

File: python/social_bot/envs/pr2.py [hunk 3]
```diff
@@ -58,128 +81,288 @@ def __init__(self,
...
-        self._resized_image_size = (84, 84)
-
-        self._prev_dist = self._get_finger_tip_distance()
-        self._collision_cnt = 0
+        self._resized_image_size = (128, 128) #(84, 84)
+        self._use_internal_states_only = use_internal_states_only
         self._cum_reward = 0.0
+        self._motion_loss = motion_loss
+
+        # a hack to walk around gripper not open initially, might not need now.
```
[the remark this answers is not in the document]
Comment (bayesian) [reply python/social_bot/envs/pr2.py]: yes

File: python/social_bot/envs/pr2.py [hunk 4]
```diff
@@ -58,128 +81,288 @@ def __init__(self,
...
+            obs = np.concatenate((
+                np.array(joint_positions), np.array(joint_velocities), np.array(loc),
+                np.array(loc_a), np.array(l_finger_tip_loc), np.array(l_finger_tip_loc_a),
+                np.array(r_finger_tip_loc), np.array(r_finger_tip_loc_a),
+                np.array([l_dist, r_dist]),
+                np.array([self._l_touch, self._r_touch]).astype(np.float32)), 0)
+        else:
+            obs = np.array(
+                self._agent.get_camera_observation(
+                    "default::pr2::pr2::head_tilt_link::head_mount_prosilica_link_sensor"
```
[the remark this answers is not in the document]
Comment (bayesian) [reply python/social_bot/envs/pr2.py]: agree, currently only the internal states observation works, we shall investigate into stereo camera later.

File: python/social_bot/envs/pr2.py (hunk 1 above)
Comment (bayesian) [reply python/social_bot/envs/pr2.py]: yes, ||f||_2 could be better here, but we did not use it yet.

File: python/social_bot/envs/pr2.py [hunk 5]
```diff
@@ -58,128 +81,288 @@ def __init__(self,
...
+
+        # final actions used to control each joints are:
+        # joint_limit_{i} *  \pi (state)_{i}
+        actions = scaled_actions
+
+        controls = dict(zip(self._r_arm_joints, actions))
 
         self._agent.take_action(controls)
         self._world.step(100)
+
```
[the remark this answers is not in the document]
Comment (bayesian) [reply python/social_bot/envs/pr2.py]: just follow the convention as other environments, also I don't think we need such frequency controls here.. 10-20 steps have two problems: frequency too high 50-100 hz, and also episode and simulation too slow.. 10x longer

File: python/social_bot/envs/pr2.py [hunk 6]
```diff
@@ -38,17 +38,40 @@ class Pr2Gripper(gym.Env):
...
+                 motion_loss=0.0000,
+                 use_internal_states_only=True,
                  port=None):
         """
         Args:
             goal_name (string): name of the object to lift off ground
             max_steps (int): episode will end when the agent exceeds the number of steps.
             reward_shaping (boolean): whether it adds distance based reward shaping.
+            motion_loss (float): if not zero, it will add -motion_loss * || V_{all_joints} ||^2 in
+                 the reward when episode endds.
```
Comment (emailweixu) [inline python/social_bot/envs/pr2.py]: endds => ends

File: pygazebo/pygazebo.cc [hunk 7]
```diff
@@ -440,8 +462,8 @@ PYBIND11_MODULE(pygazebo, m) {
            py::arg("sensor_scope_name"))
       .def("get_joint_state", &Agent::GetJointState, py::arg("joint_name"))
       .def("get_link_pose", &Agent::GetLinkPose, py::arg("link_name"))
-      .def("get_collision_count",
-           &Agent::GetCollisionCount,
+      .def("get_collisions",
+           &Agent::GetCollisions,
```
Comment (emailweixu) [inline pygazebo/pygazebo.cc]: Add a help message to describe the function in .def()

File: python/social_bot/envs/pr2.py [hunk 8]
```diff
@@ -58,128 +81,288 @@ def __init__(self,
...
 
         self._all_joints = self._agent.get_joint_names()
 
-        self._world.info()
+        # to avoid different parallel simulation has the same randomness
+        random.seed(port)
+
+        #self._world.info()
+
+        # from https://github.com/PR2/pr2_mechanism/blob/melodic-devel/pr2_mechanism_model/pr2.urdf
```
Comment (emailweixu) [inline python/social_bot/envs/pr2.py]: comment what a passive joint is

File: python/social_bot/envs/pr2.py [hunk 9]
```diff
@@ -58,128 +81,288 @@ def __init__(self,
...
+        gripper_pos = self._get_gripper_pos()
 
         done = self._steps_in_this_episode >= self._max_steps
 
-        reward = -0.01 if (not done) else (1.2 - dist)
+        reward = 0 if (not done) else 1.0 - np.tanh(5.0 * dist)
+
+        dist_reward = self._prev_dist - dist
+
+        pos_reward = self._gripper_reward_dir * (gripper_pos - self._prev_gripper_pos)
```
Comment (emailweixu) [inline python/social_bot/envs/pr2.py]: Give a description about this reward

File: python/social_bot/envs/pr2.py [hunk 10]
```diff
@@ -58,128 +81,288 @@ def __init__(self,
...
+        r_finger_tip_loc = np.array(self._r_finger_pose[0])
 
-        dist = np.linalg.norm(np.array(finger_tip_loc) - np.array(goal_loc))
+        finger_tip_center = 0.5 * (l_finger_tip_loc + r_finger_tip_loc)
+        dist = np.linalg.norm(np.array(finger_tip_center) - np.array(goal_loc))
         return dist
 
-    def step(self, action):
-        controls = dict(zip(self._r_arm_joints, action))
+    def _get_gripper_pos(self):
```
Comment (emailweixu) [inline python/social_bot/envs/pr2.py]: comment that this is the angle

File: python/social_bot/envs/pr2.py [hunk 11]
```diff
@@ -58,128 +81,288 @@ def __init__(self,
...
+        state = self._agent.get_joint_state("pr2::pr2::r_gripper_joint")
+        return state.get_positions()[0]
+
+    def step(self, actions):
+        def trunc_scale(a, limit):
```
Comment (emailweixu) [inline python/social_bot/envs/pr2.py]: min(max(a, -1), 1) * limit

File: python/social_bot/envs/pr2.py [hunk 12]
```diff
@@ -38,17 +38,40 @@ class Pr2Gripper(gym.Env):
     currently, the agent only learns to use arm/hand to play bat to hit the goal
```
Comment (emailweixu) [inline python/social_bot/envs/pr2.py]: And describe how the reward is given.

File: python/social_bot/envs/pr2.py [hunk 13]
```diff
@@ -58,128 +81,288 @@ def __init__(self,
...
+                                           "beer::link::box_collision")
+        self._goal_pose = self._goal.get_pose()
+        self._l_finger_pose = self._agent.get_link_pose("pr2::pr2::r_gripper_l_finger_tip_link")
+        self._r_finger_pose = self._agent.get_link_pose("pr2::pr2::r_gripper_r_finger_tip_link")
+
+        if self._use_internal_states_only:
+            joint_states = [self._agent.get_joint_state(joint) for joint in self._r_arm_joints]
+            joint_positions = [s.get_positions()[0] for s in joint_states]
+            joint_velocities = [s.get_velocities()[0] for s in joint_states]
+            loc, loc_a = self._goal_pose
```
Comment (emailweixu) [inline python/social_bot/envs/pr2.py]: should consider using logical camera to get the relative pose of the objects in the view

File: examples/train_pr2.py [hunk 14]
```diff
@@ -32,5 +41,7 @@ def eval_ppo2():
 
 
 if __name__ == '__main__':
+    config = tf.ConfigProto()
+    config.gpu_options.allow_growth = True
     logging.basicConfig(level=logging.DEBUG)
-    ppo2_lstm()
+    ppo2_mlp()
```
Comment (Jialn) [inline examples/train_pr2.py]: Is that from baseline? I remember you said that tf-agent ppo was used and many parameters are tuned.

File: python/social_bot/envs/pr2.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Jialn) [reply python/social_bot/envs/pr2.py]: the max_step_size option in world file should be higher to keep time scale the same. 
i.e., max_step_size 0.001 -> 0.005,  if 100substeps -> 20 substeps. Control frequency won't change here, but the simulator will be 5x faster.

File: examples/train_pr2.py (hunk 14 above)
Comment (bayesian) [reply examples/train_pr2.py]: it is in alf depot, will send out PR shortly.

File: python/social_bot/envs/pr2.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (emailweixu) [reply python/social_bot/envs/pr2.py]: Cannot understand your response

File: pygazebo/pygazebo.cc (hunk 7 above)
Comment (emailweixu) [reply pygazebo/pygazebo.cc]: Can you also add a help message for this function?

File: python/social_bot/envs/pr2.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bayesian) [reply python/social_bot/envs/pr2.py]: I will try that when train with camera based observation

File: pygazebo/pygazebo.cc (hunk 7 above)
Comment (bayesian) [reply pygazebo/pygazebo.cc]: done

File: pygazebo/pygazebo.cc [hunk 15]
```diff
@@ -166,11 +177,11 @@ class Agent : public Model {
     }
   }
 
-  // TODO: this is not the correct way to expose contacts from gazebo
-  // we could expose api to check whether two links are in contact,
-  // the other way of registering async callback does not seem to fit
-  // in the current pygazebo interface.
-  unsigned int GetCollisionCount(const std::string& contact_sensor_name) {
+  std::set<std::tuple<std::string, std::string>> GetCollisions(
```
[the remark this answers is not in the document]
Comment (bayesian) [reply pygazebo/pygazebo.cc]: done
